### ISSUE TITLE
Offer a way to rewrite scalaJS imports at linktime

### ISSUE THREAD

**Issue by Quafadas** [VirtusLab/scala-cli#2698]
Proposal: Offer a way to rewrite scalaJS imports at linktime. 

Here's the prior art
https://github.com/armanbilge/scalajs-importmap

Some prior discussions for context
https://github.com/VirtusLab/scala-cli/discussions/1968#discussioncomment-5446977

https://github.com/VirtusLab/scala-cli/issues/297#issuecomment-1485161038

My goal is to see if I get agreement on 
1. This being desirable in scala-cli (I think it is) - i.e. that a PR turning the CI green with this would get merged
2. Agreement on a solution sketch

My aim, would be to be able to write something like;

project.scala
```
//> using scala 3.3.1
//> using platform js
//> using jsImportMap @stdlib/blas/base=https://cdn.jsdelivr.net/gh/stdlib-js/blas@esm/index.mjs
```

facade.scala
```
import scala.scalajs.js
import scala.scalajs.js.annotation.JSImport
import scala.scalajs.js.typedarray.Float64Array

@js.native
@JSImport("@stdlib/blas/base", JSImport.Default)
object blas extends BlasArrayOps

@js.native
trait BlasArrayOps extends js.Object:
    def dnrm2(N: Int, x: Float64Array, strideX: Int): Double = js.native
end BlasArrayOps

@main def run =
  println(blas.dnrm2(3,Float64Array(1.0, 2.0, 3.0), 0).mkString(","))
```
The expected behaviour, would be for scala-cli to rewrite the import at link time, and rely on the JS module system to resolve the dependancy. Hopefully, this sweeps away all of the bundling / build tool complexity of leveraging the wider JS ecosystem.

I intend to try and PR it myself, but I would doubt I can do it without (a lot of) help. My concrete aim in this issue is for the two top points to be agreed and clearly laid out, so that others have a chance to follow the why and how breadcrumbs, too, should they wish. 

Here's my best stab at conretely, what I think would need to happen. 

Firstly, Introduce https://github.com/armanbilge/scalajs-importmap as a dependancy to scala-cli. Then, we need to be able to feed it what it is, that we want to remap. The most "scala-cli" like approach, would appear to be to lean into the using directives. i.e. add a jsImportMap directive here?
https://github.com/VirtusLab/scala-cli/blob/main/modules/cli/src/main/scala/scala/cli/commands/shared/ScalaJsOptions.scala

Then, need to figure out where in scala-cli, we have access to the JS IR (<-- help!), then wheel in Armans library and the values set in the new directive. 

Then... unit test (where? <--- help!) and iterate...

**Comment by Quafadas** [comment]
@armanbilge Do you think this is a reasonable description?

**Comment by armanbilge** [comment]
```
//> using jsImportMap
```

One idea I have for this is to pass a path to a file that follows the same Import map JSON representation that is supported by browsers.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap#import_map_json_representation

---

> Then, need to figure out where in scala-cli, we have access to the JS IR (<-- help!),

Actually, this doesn't happen in scala-cli at all. You want to make modifications here:

https://github.com/VirtusLab/scala-js-cli/

That is a CLI interface to scala-js. You will want to wheel in the scalajs-importmap library there and expose it as a CLI option.

Then, in scala-cli, you would change how that scala-js-cli is invoked from scala-cli i.e. translate the directives to CLI flags.

**Comment by Quafadas** [comment]
Ohhh, thanks... Attractive about this would seem to me to be, that it ought to be possible to evolve https://github.com/VirtusLab/scala-js-cli/ independently of the larger scala-cli project, I hope in a compatible way, in a first and low risk step.

Before I gamely plow on; the current dependancy list of scala-js-cli very short. From what I can tell, it does not include things for file reading / JSON. In one of my own projects, I'd add in os-lib and upickle. Would that be problematic in a project which is certainly upstream from those? I hope, that the CLI nature of it, means it's published independently and there's no class path / version conflicts. But checking to see if that would be problematic to add those libs as dependancies to scala-js-cli?

**Comment by Gedochao** [comment]
> Before I gamely plow on; the current dependancy list of scala-js-cli very short. From what I can tell, it does not include things for file reading / JSON. In one of my own projects, I'd add in os-lib and upickle. Would that be problematic in a project which is certainly upstream from those? I hope, that the CLI nature of it, means it's published independently and there's no class path / version conflicts. But checking to see if that would be problematic to add those libs as dependancies to scala-js-cli?

`os-lib` is used by `scala-cli` as well, so that's fine.
For JSON stuff we use [jsoniter-scala](https://github.com/plokhotnyuk/jsoniter-scala), so perhaps that'd be better for consistency... Although if it turns out to be overkill, I think `upickle` would still be okay.

**Comment by Gedochao** [comment]
> Then... unit test (where? <--- help!) and iterate...

This trait is where (most) our Scala.js-related (integration) tests are:

https://github.com/VirtusLab/scala-cli/blob/main/modules/integration/src/test/scala/scala/cli/integration/RunScalaJsTestDefinitions.scala

The few related unit tests would be around here:
https://github.com/VirtusLab/scala-cli/blob/702d925e278953fb9f7018b0e03b576c256c7f57/modules/build/src/test/scala/scala/build/tests/BuildTests.scala#L986

**Comment by Quafadas** [comment]
A start, for step 1. 
https://github.com/VirtusLab/scala-js-cli/pull/47

**Comment by Gedochao** [comment]
Note: https://github.com/VirtusLab/scala-js-cli/pull/47 has been merged and is currently available in [`scala-js-cli` v1.15.0.1](https://github.com/VirtusLab/scala-js-cli/releases/tag/v1.15.0.1)
